intermediate_source/pipeline_tutorial.py

```diff
@@ -0,0 +1,415 @@
+# ``PositionalEncoding`` module injects some information about the
+# relative or absolute position of the tokens in the sequence. The
+# positional encodings have the same dimension as the embeddings so that
+# the two can be summed. Here, we use ``sine`` and ``cosine`` functions of
+# different frequencies.
+
+
+class PositionalEncoding(nn.Module):
+
+    def __init__(self, d_model, dropout=0.1, max_len=5000):
+        super(PositionalEncoding, self).__init__()
+        self.dropout = nn.Dropout(p=dropout)
+
+        pe = torch.zeros(max_len, d_model)
+        position = torch.arange(0, max_len, dtype=torch.float).unsqueeze(1)
+        div_term = torch.exp(torch.arange(0, d_model, 2).float() * (-math.log(10000.0) / d_model))
+        pe[:, 0::2] = torch.sin(position * div_term)
+        pe[:, 1::2] = torch.cos(position * div_term)
+        pe = pe.unsqueeze(0).transpose(0, 1)
+        self.register_buffer('pe', pe)
+
+    def forward(self, x):
+        x = x + self.pe[:x.size(0), :]
+        return self.dropout(x)
+
+
+
+######################################################################
+# Load and batch data
+# -------------------
+#
+
+
+######################################################################
+# The training process uses Wikitext-2 dataset from ``torchtext``. The
+# vocab object is built based on the train dataset and is used to numericalize
+# tokens into tensors. Starting from sequential data, the ``batchify()``
+# function arranges the dataset into columns, trimming off any tokens remaining
+# after the data has been divided into batches of size ``batch_size``.
+# For instance, with the alphabet as the sequence (total length of 26)
+# and a batch size of 4, we would divide the alphabet into 4 sequences of
+# length 6:
+#
+# .. math::
+#   \begin{bmatrix}
+#   \text{A} & \text{B} & \text{C} & \ldots & \text{X} & \text{Y} & \text{Z}
+#   \end{bmatrix}
+#   \Rightarrow
+#   \begin{bmatrix}
+#   \begin{bmatrix}\text{A} \\ \text{B} \\ \text{C} \\ \text{D} \\ \text{E} \\ \text{F}\end{bmatrix} &
+#   \begin{bmatrix}\text{G} \\ \text{H} \\ \text{I} \\ \text{J} \\ \text{K} \\ \text{L}\end{bmatrix} &
+#   \begin{bmatrix}\text{M} \\ \text{N} \\ \text{O} \\ \text{P} \\ \text{Q} \\ \text{R}\end{bmatrix} &
+#   \begin{bmatrix}\text{S} \\ \text{T} \\ \text{U} \\ \text{V} \\ \text{W} \\ \text{X}\end{bmatrix}
+#   \end{bmatrix}
+#
+# These columns are treated as independent by the model, which means that
+# the dependence of ``G`` and ``F`` can not be learned, but allows more
+# efficient batch processing.
+#
+
+import io
+import torch
+from torchtext.utils import download_from_url, extract_archive
+from torchtext.data.utils import get_tokenizer
+from torchtext.vocab import build_vocab_from_iterator
+
+url = 'https://s3.amazonaws.com/research.metamind.io/wikitext/wikitext-2-v1.zip'
+test_filepath, valid_filepath, train_filepath = extract_archive(download_from_url(url))
```

Comment:
You need to create `.data` dir here, or it fails.

Reply:
Hmm, this did work for me without creating `.data`, the directory was created as part of the download.

Reply:
```
python -c "from torchtext.utils import download_from_url, extract_archive; extract_archive(download_from_url('https://s3.amazonaws.com/research.metamind.io/wikitext/wikitext-2-v1.zip'))"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/stas/anaconda3/envs/main-38/lib/python3.8/site-packages/torchtext/utils.py", line 82, in download_from_url
    raise RuntimeError(
RuntimeError: Download directory .data does not exist. Did you create it?
```
```
python -c "import torchtext; print(torchtext.__version__)"
0.6.0
```

Reply:
update: it works just fine if I upgrade it to `torchtext==0.9.0`